src/SessionSetup/ConnectToTargetDialog.cpp

```diff
@@ -31,160 +32,143 @@ ConnectToTargetDialog::ConnectToTargetDialog(
+}
+
+ErrorMessageOr<orbit_session_setup::ServiceDeployManager::GrpcPort>
+ConnectToTargetDialog::DeployOrbitService(
+    orbit_session_setup::ServiceDeployManager* service_deploy_manager) {
+  orbit_ssh_qt::ScopedConnection label_connection{
+      QObject::connect(service_deploy_manager, &ServiceDeployManager::statusMessage, this,
+                       &ConnectToTargetDialog::SetStatusMessage)};
   orbit_ssh_qt::ScopedConnection cancel_connection{
-      QObject::connect(ui_->abortButton, &QPushButton::clicked,
-                       result_data_.service_deploy_manager_.get(), &ServiceDeployManager::Cancel)};
+      QObject::connect(ui_->abortButton, &QPushButton::clicked, service_deploy_manager,
+                       &ServiceDeployManager::Cancel)};
 
-  auto deployment_result = result_data_.service_deploy_manager_->Exec();
+  auto deployment_result = service_deploy_manager->Exec();
   if (deployment_result.has_error()) {
-    result_data_.error_message_ = QString::fromStdString(deployment_result.error().message());
-    return std::nullopt;
+    return ErrorMessage{"Error during service deployment"};
   } else {
     return deployment_result.value();
   }
 }
 
-void ConnectToTargetDialog::ListProcessesAndSaveProcess(
-    std::shared_ptr<grpc_impl::Channel> grpc_channel) {
+ErrorMessageOr<std::unique_ptr<orbit_client_data::ProcessData>>
+ConnectToTargetDialog::FindSpecifiedProcess(std::shared_ptr<grpc_impl::Channel> grpc_channel,
+                                            uint32_t process_id) {
   CHECK(grpc_channel != nullptr);
 
   orbit_client_services::ProcessClient client(grpc_channel);
   auto process_list = client.GetProcessList();
-  if (!process_list.has_value()) return;
+  if (!process_list.has_value()) {
+    return ErrorMessage("Unknown error: Could not retrieve list of running processes.");
+  }
 
   for (auto& process : process_list.value()) {
-    if (process.pid() == process_id_) {
-      result_data_.process_data_ = std::make_unique<orbit_client_data::ProcessData>(process);
-      return;
+    if (process.pid() == process_id) {
+      return std::make_unique<orbit_client_data::ProcessData>(process);
     }
   }
 
-  result_data_.error_message_ =
-      QString("PID %1 was not found in the list of running processes.").arg(process_id_);
-}
-
-void ConnectToTargetDialog::CloseDialogIfResultIsReady() {
-  if (!result_data_.error_message_.isEmpty()) {
-    reject();
-  }
-
-  if (result_data_.maybe_instance_.has_value() && result_data_.grpc_channel_ != nullptr &&
-      result_data_.process_data_ != nullptr) {
-    accept();
-  }
+  return ErrorMessage(QString("PID %1 was not found in the list of running processes.")
+                          .arg(process_id)
+                          .toStdString());
 }
 
 void ConnectToTargetDialog::SetStatusMessage(const QString& message) {
   ui_->statusLabel->setText(QString("<b>Status:</b> ") + message);
 }
 
+void ConnectToTargetDialog::LogAndDisplayError(const QString& message) {
```

Comment:
pers pref. This is always called with `ErrorMessage`.  Doing the conversion to QString in here, saves a few words